File: Wayfork/WayforkCore/Sources/WayforkCore/App/ImportExport.swift
import Foundation

public enum ImportMode: Sendable, Hashable {
    case replace
    case merge
}

public struct ImportPreview: Sendable, Hashable {
    public var tunnels: Int
    public var rules: Int
    public var includesSecrets: Bool
    public var tunnelsWithSecrets: Int

    public init(tunnels: Int, rules: Int, includesSecrets: Bool, tunnelsWithSecrets: Int) {
        self.tunnels = tunnels
        self.rules = rules
        self.includesSecrets = includesSecrets
        self.tunnelsWithSecrets = tunnelsWithSecrets
    }
}

public struct ImportOutcome: Sendable, Equatable {
    public var store: Store
    public var secrets: [SecretKey: String]
    public var warnings: [String]
    public var tunnelsAdded: Int
    public var tunnelsUpdated: Int
    public var rulesAdded: Int
    public var rulesUpdated: Int
    public var rulesSkipped: Int
}

public enum StoreImporter {
    public static func preview(_ document: ExportDocument) -> ImportPreview {
        ImportPreview(
            tunnels: document.tunnels.count,
            rules: document.rules.count,
            includesSecrets: document.includesSecrets,
            tunnelsWithSecrets: document.tunnels.count { !$0.secrets.isEmpty })
    }

    public static func apply(
        _ document: ExportDocument, to store: Store, mode: ImportMode
    ) -> ImportOutcome {
        var result =
            mode == .replace
            ? Store(schemaVersion: store.schemaVersion, settings: document.settings)
            : store
        var secrets: [SecretKey: String] = [:]
        var warnings: [String] = []
        var tunnelsAdded = 0
        var tunnelsUpdated = 0

        for exported in document.tunnels {
            if mode == .merge,
                let index = result.tunnels.firstIndex(where: { $0.id == exported.id })
            {
                let name = availableName(
                    exported.name, for: exported.id, in: result, warnings: &warnings)
                result.tunnels[index].name = name
                result.tunnels[index].isEnabled = exported.isEnabled
                result.tunnels[index].kind = exported.kind
                result.tunnels[index].createdAt = exported.createdAt
                tunnelsUpdated += 1
                collect(exported.secrets, for: exported.id, into: &secrets)
                continue
            }

            guard let slot = result.nextFreeSlot() else {
                warnings.append("Tunnel \(exported.name) skipped: no free slots")
                continue
            }
            let name = availableName(
                exported.name, for: exported.id, in: result, warnings: &warnings)
            result.tunnels.append(
                Tunnel(
                    id: exported.id,
                    name: name,
                    isEnabled: exported.isEnabled,
                    slot: slot,
                    kind: exported.kind,
                    createdAt: exported.createdAt))
            tunnelsAdded += 1
            collect(exported.secrets, for: exported.id, into: &secrets)
        }

        var rulesAdded = 0
        var rulesUpdated = 0
        var rulesSkipped = 0
        if mode == .replace {
            result.rules = []
        }
        let tunnelIDs = Set(result.tunnels.map(\.id))
        for rule in document.rules {
            if let tunnelID = rule.tunnelID, !tunnelIDs.contains(tunnelID) {
                let shortID = tunnelID.uuidString.lowercased().prefix(4)
                warnings.append("Rule \(rule.pattern) skipped: tunnel \(shortID)… not found")
                rulesSkipped += 1
                continue
            }
            if mode == .merge,
                let index = result.rules.firstIndex(where: { $0.id == rule.id })
            {
                result.rules[index] = rule
                rulesUpdated += 1
            } else {
                result.rules.append(rule)
                rulesAdded += 1
            }
        }

        // F8: the file's default replaces the current one only when it names a tunnel that
        // made it in; on Replace the fresh store has none to begin with.
        if let wanted = document.defaultTunnelID {
            if tunnelIDs.contains(wanted) {
                result.defaultTunnelID = wanted
            } else {
                let shortID = wanted.uuidString.lowercased().prefix(4)
                warnings.append(
                    "Default tunnel \(shortID)… not found: everything else stays direct")
            }
        }

        return ImportOutcome(
            store: result,
            secrets: secrets,
            warnings: warnings,
            tunnelsAdded: tunnelsAdded,
            tunnelsUpdated: tunnelsUpdated,
            rulesAdded: rulesAdded,
            rulesUpdated: rulesUpdated,
            rulesSkipped: rulesSkipped)
    }

    private static func availableName(
        _ requested: String, for id: UUID, in store: Store, warnings: inout [String]
    ) -> String {
        guard !store.isNameAvailable(requested, excluding: id) else { return requested }
        var number = 2
        while true {
            let suffix = " (\(number))"
            let base = String(requested.prefix(Tunnel.nameMaxLength - suffix.count))
            let candidate = base + suffix
            if store.isNameAvailable(candidate, excluding: id) {
                warnings.append("Tunnel \(requested) renamed to \(candidate)")
                return candidate
            }
            number += 1
        }
    }

    private static func collect(
        _ tunnelSecrets: TunnelSecrets, for id: UUID, into secrets: inout [SecretKey: String]
    ) {
        if let ovpn = tunnelSecrets.ovpn { secrets[.ovpn(id)] = ovpn }
        if let credentials = tunnelSecrets.credentials,
            let data = try? JSONCoding.compactEncoder.encode(credentials)
        {
            secrets[.credentials(id)] = String(decoding: data, as: UTF8.self)
        }
        if let passphrase = tunnelSecrets.keyPassphrase {
            secrets[.keyPassphrase(id)] = passphrase
        }
        if let uuid = tunnelSecrets.uuid { secrets[.uuid(id)] = uuid }
    }
}

public enum StoreExporter {
    public static func document(
        store: Store,
        secretStore: any SecretStore,
        includeSecrets: Bool,
        exportedAt: Date = Date()
    ) throws -> ExportDocument {
        let tunnels = try store.tunnels.map { tunnel in
            guard includeSecrets else { return ExportedTunnel(tunnel: tunnel) }
            let secrets: TunnelSecrets
            switch tunnel.kind {
            case .openVPN:
                secrets = try TunnelSecrets(
                    ovpn: secretStore.read(.ovpn(tunnel.id)),
                    credentials: secretStore.readCredentials(for: tunnel.id),
                    keyPassphrase: secretStore.read(.keyPassphrase(tunnel.id)))
            case .vless:
                secrets = try TunnelSecrets(uuid: secretStore.read(.uuid(tunnel.id)))
            }
            return ExportedTunnel(tunnel: tunnel, secrets: secrets)
        }
        return ExportDocument(
            exportedAt: exportedAt,
            includesSecrets: includeSecrets,
            tunnels: tunnels,
            rules: store.rules,
            settings: store.settings,
            defaultTunnelID: store.defaultTunnelID)
    }
}

File: WayforkWindows/app/lib/core/app/import_export.dart
import 'package:wayfork/core/json_text.dart';
import 'package:wayfork/core/model/export_document.dart';
import 'package:wayfork/core/model/rule.dart';
import 'package:wayfork/core/model/store.dart';
import 'package:wayfork/core/model/tunnel.dart';
import 'package:wayfork/core/secrets/secret_store.dart';

enum ImportMode { replace, merge }

final class ImportPreview {
  const ImportPreview({
    required this.tunnels,
    required this.rules,
    required this.includesSecrets,
    required this.tunnelsWithSecrets,
  });

  final int tunnels;
  final int rules;
  final bool includesSecrets;
  final int tunnelsWithSecrets;

  @override
  bool operator ==(Object other) =>
      other is ImportPreview &&
      tunnels == other.tunnels &&
      rules == other.rules &&
      includesSecrets == other.includesSecrets &&
      tunnelsWithSecrets == other.tunnelsWithSecrets;

  @override
  int get hashCode =>
      Object.hash(tunnels, rules, includesSecrets, tunnelsWithSecrets);
}

final class ImportOutcome {
  const ImportOutcome({
    required this.store,
    required this.secrets,
    required this.warnings,
    required this.tunnelsAdded,
    required this.tunnelsUpdated,
    required this.rulesAdded,
    required this.rulesUpdated,
    required this.rulesSkipped,
  });

  final Store store;

  /// Secrets carried by the file, ready for `SecretStore.write`.
  final Map<SecretKey, String> secrets;
  final List<String> warnings;
  final int tunnelsAdded;
  final int tunnelsUpdated;
  final int rulesAdded;
  final int rulesUpdated;
  final int rulesSkipped;
}

/// Applies a `wayfork-export.json` to the store (docs/design/02-ux.md, F7).
abstract final class StoreImporter {
  static ImportPreview preview(ExportDocument document) => ImportPreview(
    tunnels: document.tunnels.length,
    rules: document.rules.length,
    includesSecrets: document.includesSecrets,
    tunnelsWithSecrets: document.tunnels
        .where((tunnel) => !tunnel.secrets.isEmpty)
        .length,
  );

  static ImportOutcome apply(
    ExportDocument document, {
    required Store to,
    required ImportMode mode,
  }) {
    var result = mode == ImportMode.replace
        ? Store(schemaVersion: to.schemaVersion, settings: document.settings)
        : to;
    final secrets = <SecretKey, String>{};
    final warnings = <String>[];
    var tunnelsAdded = 0;
    var tunnelsUpdated = 0;

    for (final exported in document.tunnels) {
      final tunnels = [...result.tunnels];
      final index = mode == ImportMode.merge
          ? tunnels.indexWhere((tunnel) => tunnel.id == exported.id)
          : -1;
      if (index >= 0) {
        final name = _availableName(
          exported.name,
          id: exported.id,
          store: result,
          warnings: warnings,
        );
        tunnels[index] = tunnels[index].copyWith(
          name: name,
          isEnabled: exported.isEnabled,
          kind: exported.kind,
          createdAt: exported.createdAt,
        );
        result = result.copyWith(tunnels: tunnels);
        tunnelsUpdated += 1;
        _collect(exported.secrets, id: exported.id, into: secrets);
        continue;
      }

      final slot = result.nextFreeSlot();
      if (slot == null) {
        warnings.add('Tunnel ${exported.name} skipped: no free slots');
        continue;
      }
      final name = _availableName(
        exported.name,
        id: exported.id,
        store: result,
        warnings: warnings,
      );
      tunnels.add(
        Tunnel(
          id: exported.id,
          name: name,
          isEnabled: exported.isEnabled,
          slot: slot,
          kind: exported.kind,
          createdAt: exported.createdAt,
        ),
      );
      result = result.copyWith(tunnels: tunnels);
      tunnelsAdded += 1;
      _collect(exported.secrets, id: exported.id, into: secrets);
    }

    var rulesAdded = 0;
    var rulesUpdated = 0;
    var rulesSkipped = 0;
    final rules = mode == ImportMode.replace ? <Rule>[] : [...result.rules];
    final tunnelIDs = result.tunnels.map((tunnel) => tunnel.id).toSet();
    for (final rule in document.rules) {
      final tunnelID = rule.tunnelID;
      if (tunnelID != null && !tunnelIDs.contains(tunnelID)) {
        warnings.add(
          'Rule ${rule.pattern} skipped: tunnel ${_short(tunnelID)}… not found',
        );
        rulesSkipped += 1;
        continue;
      }
      final index = mode == ImportMode.merge
          ? rules.indexWhere((existing) => existing.id == rule.id)
          : -1;
      if (index >= 0) {
        rules[index] = rule;
        rulesUpdated += 1;
      } else {
        rules.add(rule);
        rulesAdded += 1;
      }
    }
    result = result.copyWith(rules: rules);

    // F8: the file's default replaces the current one only when it names a
    // tunnel that made it in; on Replace the fresh store has none to begin with.
    final wanted = document.defaultTunnelID;
    if (wanted != null) {
      if (tunnelIDs.contains(wanted)) {
        result = result.copyWith(defaultTunnelID: wanted);
      } else {
        warnings.add(
          'Default tunnel ${_short(wanted)}… not found: everything else stays '
          'direct',
        );
      }
    }

    return ImportOutcome(
      store: result,
      secrets: secrets,
      warnings: warnings,
      tunnelsAdded: tunnelsAdded,
      tunnelsUpdated: tunnelsUpdated,
      rulesAdded: rulesAdded,
      rulesUpdated: rulesUpdated,
      rulesSkipped: rulesSkipped,
    );
  }

  static String _short(String id) => id.substring(0, 4);

  static String _availableName(
    String requested, {
    required String id,
    required Store store,
    required List<String> warnings,
  }) {
    if (store.isNameAvailable(requested, excluding: id)) return requested;
    var number = 2;
    while (true) {
      final suffix = ' ($number)';
      final keep = Tunnel.nameMaxLength - suffix.length;
      final base = requested.length > keep
          ? requested.substring(0, keep)
          : requested;
      final candidate = base + suffix;
      if (store.isNameAvailable(candidate, excluding: id)) {
        warnings.add('Tunnel $requested renamed to $candidate');
        return candidate;
      }
      number += 1;
    }
  }

  static void _collect(
    TunnelSecrets tunnelSecrets, {
    required String id,
    required Map<SecretKey, String> into,
  }) {
    final ovpn = tunnelSecrets.ovpn;
    if (ovpn != null) into[SecretKey(SecretKind.ovpn, id)] = ovpn;
    final credentials = tunnelSecrets.credentials;
    if (credentials != null) {
      into[SecretKey(SecretKind.credentials, id)] = JsonCoding.encodeCompact(
        credentials.toJson(),
      );
    }
    final passphrase = tunnelSecrets.keyPassphrase;
    if (passphrase != null) {
      into[SecretKey(SecretKind.keyPassphrase, id)] = passphrase;
    }
    final uuid = tunnelSecrets.uuid;
    if (uuid != null) into[SecretKey(SecretKind.uuid, id)] = uuid;
  }
}

abstract final class StoreExporter {
  static Future<ExportDocument> document({
    required Store store,
    required SecretStore secretStore,
    required bool includeSecrets,
    DateTime? exportedAt,
  }) async {
    final tunnels = <ExportedTunnel>[];
    for (final tunnel in store.tunnels) {
      if (!includeSecrets) {
        tunnels.add(ExportedTunnel.fromTunnel(tunnel));
        continue;
      }
      final TunnelSecrets secrets;
      switch (tunnel.kind) {
        case TunnelKindOpenVPN():
          secrets = TunnelSecrets(
            ovpn: await secretStore.read(SecretKey(SecretKind.ovpn, tunnel.id)),
            credentials: await secretStore.readCredentials(tunnel.id),
            keyPassphrase: await secretStore.read(
              SecretKey(SecretKind.keyPassphrase, tunnel.id),
            ),
          );
        case TunnelKindVLESS():
          secrets = TunnelSecrets(
            uuid: await secretStore.read(SecretKey(SecretKind.uuid, tunnel.id)),
          );
      }
      tunnels.add(ExportedTunnel.fromTunnel(tunnel, secrets: secrets));
    }
    return ExportDocument(
      exportedAt: exportedAt,
      includesSecrets: includeSecrets,
      tunnels: tunnels,
      rules: store.rules,
      settings: store.settings,
      defaultTunnelID: store.defaultTunnelID,
    );
  }
}
